hotel/src/crypto/aes.rs

```diff
@@ -0,0 +1,155 @@
+use core::cell::Cell;
+use hil::aes::{self, AesClient, Interrupt, AesModule, ParsedInterrupt};
+use hil::common::SyscallError;
+use super::keymgr::{KEYMGR0_REGS, Registers};
+
+pub struct AesEngine {
+    regs: *mut Registers,
+    client: Cell<Option<&'static AesClient>>,
+}
+
+impl AesEngine {
+    const unsafe fn new(regs: *mut Registers) -> AesEngine {
+        AesEngine {
+            regs: regs,
+            client: Cell::new(None),
+        }
+    }
+
+    pub fn set_client(&self, client: &'static AesClient) {
+        self.client.set(Some(client));
+    }
+
+    pub fn setup(&self, key_size: aes::KeySize, key: &[u8]) -> Result<isize, SyscallError> {
+        let ref regs = unsafe { &*self.regs }.aes;
+
+        self.enable_all_interrupts();
+        regs.ctrl.set(regs.ctrl.get() | key_size as u32 | AesModule::Enable as u32);
+        let mut i = 0;
+        for word in key.chunks(4) {
+            if i >= 8 {
+                println!("Input key too long!");
+                return Err(SyscallError::InvalidState);
+            }
+            let d = word.iter()
+                .map(|b| *b as u32)
+                .enumerate()
+                .fold(0, |accm, (i, byte)| accm | (byte << (i * 8)));
+            regs.key[i].set(d);
+            i += 1;
+        }
+        regs.key_start.set(1);
+
+        Ok(0)
+    }
+
+    pub fn set_encrypt_mode(&self, encrypt: bool) -> Result<isize, SyscallError> {
+        let ref regs = unsafe { &*self.regs }.aes;
+
+        let flag = aes::Mode::Encrypt as u32;
+        if encrypt {
+            regs.ctrl.set(regs.ctrl.get() | flag);
+        } else if regs.ctrl.get() & flag != 0 {
```

Comment:
Is there any special reason why you check the existing register value before setting it? You don't do it for the other if branch. You can just do an unconditional clear of the flag (`regs.ctrl.set(regs.ctrl.get() & !flag)`)